src/tvu/repair/service.zig

```diff
@@ -0,0 +1,436 @@
+    my_pubkey: Pubkey,
+    my_shred_version: *const Atomic(u16),
+
+    const RepairPeers = struct {
+        insertion_time_secs: u64,
+        peers: []RepairPeer,
+
+        fn deinit(self: *@This(), allocator: Allocator) void {
+            allocator.free(self.peers);
+        }
+    };
+
+    pub fn init(
+        allocator: Allocator,
+        rng: Random,
+        gossip: *RwMux(GossipTable),
+        my_pubkey: Pubkey,
+        my_shred_version: *const Atomic(u16),
+    ) error{OutOfMemory}!RepairPeerProvider {
+        return .{
+            .allocator = allocator,
+            .gossip = gossip,
+            .cache = try LruCacheCustom(.non_locking, Slot, RepairPeers, Allocator, RepairPeers.deinit)
+                .initWithContext(allocator, REPAIR_PEERS_CACHE_CAPACITY, allocator),
+            .my_pubkey = my_pubkey,
+            .my_shred_version = my_shred_version,
+            .rng = rng,
+        };
+    }
+
+    pub fn deinit(self: *@This()) void {
+        self.cache.deinit();
+    }
+
+    /// Selects a peer at random from gossip or cache that is expected
+    /// to be able to handle a repair request for the specified slot.
+    pub fn getRandomPeer(self: *@This(), slot: Slot) !?RepairPeer {
+        const peers = try self.getPeers(slot);
+        if (peers.len == 0) return null;
+        const index = self.rng.intRangeLessThan(usize, 0, peers.len);
+        return peers[index];
+    }
+
+    /// Tries to get peers that could have the slot. Checks cache, falling back to gossip.
+    fn getPeers(self: *@This(), slot: Slot) ![]RepairPeer {
+        const now: u64 = @intCast(std.time.timestamp());
+
+        if (self.cache.get(slot)) |peers| {
+            if (now - peers.insertion_time_secs <= REPAIR_PEERS_CACHE_TTL_SECONDS) {
+                return peers.peers;
+            }
+        }
+
+        const peers = try self.getRepairPeersFromGossip(self.allocator, slot);
+        try self.cache.insert(slot, .{
+            .insertion_time_secs = now,
+            .peers = peers,
+        });
+        return peers;
+    }
+
+    /// Gets a list of peers from the gossip table that are likely to have the desired slot.
+    /// This will always acquire the gossip table lock.
+    /// Instead of using this function, access the cache when possible to avoid contention.
+    fn getRepairPeersFromGossip(
+        self: *@This(),
+        allocator: Allocator,
+        slot: Slot,
+    ) error{OutOfMemory}![]RepairPeer {
+        var reader = self.gossip.read();
+        defer reader.unlock();
+        const gossip: *const GossipTable = reader.get();
+        const buf = try allocator.alloc(RepairPeer, gossip.contact_infos.count());
+        errdefer allocator.free(buf);
+        var i: usize = 0;
+        var infos = gossip.contactInfoIterator(0);
+        while (infos.next()) |info| {
+            const socket = info.getSocket(sig.gossip.socket_tag.SERVE_REPAIR);
```

Comment:
nit: change to `repair_socket`

Reply:
I'm calling it "serve_repair_socket" instead since that's a different socket.

all renames in 1 commit: https://github.com/Syndica/sig/pull/87/commits/4e02024cfbff9143db5fdf9ec08dbb32a2cef83a